testgres/pubsub.py

```diff
@@ -0,0 +1,200 @@
+            node: subscriber's node
+            publication: :class:`.Publication` object we are subscribing to
+                (see :meth:`.PostgresNode.publish()`)
+            dbname: database name used to connect and perform subscription
+            username: username used to connect to the database
+            params: subscription parameters (see documentation on `CREATE SUBSCRIPTION
+                 <https://www.postgresql.org/docs/current/static/sql-createsubscription.html>`_
+                 for details)
+        """
+        self.name = name
+        self.node = node
+        self.pub = publication
+
+        # connection info
+        conninfo = {
+            "dbname": self.pub.dbname,
+            "user": self.pub.username,
+            "host": self.pub.node.host,
+            "port": self.pub.node.port
+        }
+
+        query = (
+            "create subscription {} connection '{}' publication {}").format(
+                name, options_string(**conninfo), self.pub.name)
+
+        # additional parameters
+        if params:
+            query += " with ({})".format(options_string(**params))
+
+        node.safe_psql(query, dbname=dbname, username=username)
+
+    def disable(self, dbname=None, username=None):
+        """
+        Disables the running subscription.
+        """
+        query = "alter subscription {} disable"
+        self.node.safe_psql(query.format(self.name), dbname=None, username=None)
+
+    def enable(self, dbname=None, username=None):
+        """
+        Enables the previously disabled subscription.
+        """
+        query = "alter subscription {} enable"
+        self.node.safe_psql(query.format(self.name), dbname=None, username=None)
+
+    def refresh(self, copy_data=True, dbname=None, username=None):
+        """
+        Disables the running subscription.
+        """
+        query = "alter subscription {} refresh publication with (copy_data={})"
+        self.node.safe_psql(
+            query.format(self.name, copy_data),
+            dbname=dbname,
+            username=username)
+
+    def drop(self, dbname=None, username=None):
+        """
+        Drops subscription
+        """
+        self.node.safe_psql(
+            "drop subscription {}".format(self.name),
+            dbname=dbname,
+            username=username)
+
+    def catchup(self, username=None):
+        """
+        Wait until subscription catches up with publication.
+
+        Args:
+            username: remote node's user name
+        """
+        query = (
+            "select pg_current_wal_lsn() - replay_lsn = 0 "
+            "from pg_stat_replication where application_name = '{}'").format(
+                self.name)
+
+        try:
+            # wait until this LSN reaches subscriber
+            self.pub.node.poll_query_until(
+                query=query,
+                dbname=self.pub.dbname,
+                username=username or self.pub.username,
+                max_attempts=60)
```

Comment:
Personally, I don't like the hard-coded number.

Reply:
I think it should be a parameter here.

Reply:
Ok, added `LOGICAL_REPL_MAX_CATCHUP_ATTEMPTS` parameter